HPCC-25002 ESDL Transform Script enhancements

afishbeck opened the pull request:
1. Refactor away from ptree result handling.. allowing access to
any part of the script context.  Preserving order.  And allowing
more powerful xpath processing.
2. Refactor away from PTree based script parsing / loading.  Preserving
order.
3. Support changing the target and source locations on the fly.
4. Support builing of structured output.
5. Support calling other web services from within scripts.
6. Support dynamically tokenizing strings.  Iterating over order
accessing the result.
7. Support building structured variables.
8. Support dynamically adding namespaces to target content.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeckln committed:
HPCC-25002 ESDL Transform Script enhancements

1. Refactor away from ptree result handling.. allowing access to
   any part of the script context.  Preserving order.  And allowing
   more powerful xpath processing.
2. Refactor away from PTree based script parsing / loading.
   Preserving order.
3. Support changing the target and source locations on the fly.
4. Support builing of structured output.
5. Support calling other web services from within scripts.
6. Support dynamically tokenizing strings.  Iterating over order
   accessing the result.
7. Support building structured variables.
8. Support dynamically adding namespaces to target content.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

afishbeck commented:
@asselitx @timothyklemm please test and review.  Leaving as a draft for now, until it's functionality has been tested by someone besides me.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 4b40ab69d1b0c78dc49093cd81e8a5088bc745ea
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-232.ca-central-1.compute.internal

afishbeck commented:
Examples of usage can be found in "testing/unittests/esdltests.cpp" and in "initfiles/examples/EsdlExample/esdl_binding_entrypoints.xml".

Notice that besides examples of functionality esdl_binding_entrypoints.xml shows how order is now preserved if the script content is added as CDATA.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-232.ca-central-1.compute.internal
Sha: 4b40ab69d1b0c78dc49093cd81e8a5088bc745ea
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |148 |147 |0 |1 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Unittest error(s):
```
Run:  24   Failures: 0   Errors: 1
1) test: ESDLTests::testScriptMap (E) 
uncaught exception of unknown type

```

Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |40 sec  (00:00:40) |
|setup (roxie) |11 |11 |0 |14 sec  (00:00:14) |
|test (hthor) |936 |936 |0 |305 sec  (00:05:05) |
|test (thor) |846 |846 |0 |656 sec  (00:10:56) |
|test (roxie) |1015 |1015 |0 |351 sec  (00:05:51) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  601 sec (00:10:01) |  105 sec (00:01:45) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1628 sec (00:27:08) |   17 sec (00:00:17) | 2396 sec (00:39:56) |

timothyklemm commented on testing/unittests/esdltests.cpp:
```diff
@@ -1527,9 +1551,540 @@ constexpr const char * result = R"!!(<soap:Envelope xmlns:soap="http://schemas.x
+
+        static constexpr const char * script = R"!!(<es:CustomRequestTransform xmlns:es="urn:hpcc:esdl:script" target="Person">
+            <es:variable name='value' select="'abc'"/>
+            <es:http-post-xml url="'http://127.0.0.1:9876'" section="logging" name="roxiestuff">
```
Is it valid to assume this service is published on the localhost every time the test is run?

afishbeck replied:
I don't currently include this test in the run.  It's there for Terrence and I, and to be run explicitly for test when set up correctly.  I'd like to get to the point where we have a way to run this type of test.

timothyklemm commented on testing/unittests/esdltests.cpp:
```diff
@@ -1527,9 +1551,540 @@ constexpr const char * result = R"!!(<soap:Envelope xmlns:soap="http://schemas.x
+
+      bool legacy = false;
+      Owned<IEsdlTransformMethodMap> map = createEsdlTransformMethodMap();
+      map->addMethodTransforms("", serverScripts, legacy);
+      map->addMethodTransforms("mymethod", methodScripts, legacy);
+
+      IEsdlTransformSet *serviceSet = map->queryMethodEntryPoint("", "BackendResponse");
+      IEsdlTransformSet *methodSet = map->queryMethodEntryPoint("mymethod", "BackendResponse");
```
Just wondering if it would make sense to combine the service and method collections internally (with service transforms shared instead of duplicated) between methods? It could streamline lookup and use of transform sets by decoupling that code from knowledge about how they're defined.

afishbeck replied:
Yes it probably would but I'd have to think about the design.  I doubt I'll get that into this PR.  I may open a JIRA.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -947,74 +1490,44 @@ class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
-        m_name.set(tree.queryProp("@name"));
-        m_target.set(tree.queryProp("@target"));
+        m_name.set(stag.getValue("name"));
+        m_target.set(stag.getValue("target"));
```
Since we can use @target as a shortcut in place of a target element, and we have to for legacy, does it make sense to provide a corresponding shortcut for the source document?

afishbeck replied:
Yes. I had thought about that, and it's a good idea.  Not sure if it will make it in this PR or not, but if not I'll open a JIRA.

afishbeck replied:
done.

asselitx commented:
While running tests on the code, I see some possibly undesirable behavior that might be new. @timothyklemm I seem to recall you may have made a change to prevent this from happening. For context I'm testing a simple example with one method-level transform.

When sending the request to the back end (Roxie) the transform script is included at the location Context/Row/Common/ESP/Config/Method/Scripts

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -18,15 +18,32 @@
 {
     virtual const char *queryMergedTarget() = 0;
-    virtual bool process(IEsdlScriptContext * context, IPropertyTree *content, IXpathContext * xpathContext) = 0;
+    virtual bool process(IEsdlScriptContext * scriptContext, IXpathContext * targetContext, IXpathContext * xpathContext) = 0;
```
Definitely agree.  I'll see if I can get it changed this week.  If not I don't want to hold this PR up and I'll open a JIRA as something to clean up later.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -75,14 +92,12 @@ class CEsdlTransformOperationBase : public CInterfaceOf<IEsdlTransformOperation>
-        m_ignoreCodingErrors = tree->getPropBool("@optional", false);
+        m_tagname.set(stag.getLocalName());
+        m_traceName.set(stag.getValue("trace"));
+        m_mergedTarget.set(stag.getValue("_crtTarget"));
```
True, I'll remove it.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -91,22 +106,75 @@ class CEsdlTransformOperationBase : public CInterfaceOf<IEsdlTransformOperation>
+public:
+    CEsdlTransformOperationWithoutChildren(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationBase(xpp, stag, prefix)
+    {
+        xpp.skipSubTree();
```
I'll consider errors or exceptions.  Agree it's better to prevent them for now.

afishbeck commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +392,374 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+            throw MakeStringException(XPATHERR_InvalidState,"XpathContext:remove: ambiguous xpath xpath '%s' to remove all set 'all'", xpath);
+        }
+        xmlNodePtr *nodes = obj->nodesetval->nodeTab;
+        for (int i = 0; i < obj->nodesetval->nodeNr; i++)
```
Good catch.  I think your right, but I'll verify to make sure.

afishbeck commented:
@asselitx I'm not a huge fan of how that information gets added to the Roxie request.  I think it should be more explicit.. like add the gateway info and maybe something else.  Then if there is more needed we can do it in script.  Fixing it completely is beyond the scope of this PR.

BUT I thought I removed the script stuff explicitly and may have just missed a scenario.  Can you provide an example of the xpath into that content all the way to the start of the transforms.  I probably just need the tag names and the xpath will provide that.  I fixed it for the xpath I was using but perhaps I missed something or there is an xpath difference.

timothyklemm commented:
@afishbeck, @asselitx I think the inclusion of the Method element in the request predates the inclusion of transforms within it. I did find documentation referring to this element as a location where the roxie could look for information like gateways. It's not where our internal roxies look, but there is a precedent for inclusion. Adding **contextRemove="."** to each Method element is the solution to exclude, with other xpath options available for more granular control.

It might be worth defining a service-level value that eliminates the need to add this to every method. I'm not sure of a good way to reverse the logic and make people that actually want to see the content request it.

asselitx commented:
Yes @timothyklemm you are right. I tested with the contextRemove="." attribute and it eliminates that portion of the binding config from the context sent. For our purposes there is no need to make any immediate changes.

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -919,15 +964,29 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
+        IEsdlTransformSet *servicePLTs = m_transforms->queryMethodEntryPoint("", ESDLScriptEntryPoint_PreLogging);
+        IEsdlTransformSet *methodPLTs = m_transforms->queryMethodEntryPoint(mthdef.queryName(), ESDLScriptEntryPoint_PreLogging);
+
+        scriptContext->appendContent(ESDLScriptCtxSection_LogData, "LogDataset", logdata);
```
I think the second parameter may be misspelled. Current scripts expect the element to be named "LogDatasets". Although it is redundant the current structure is UpdateLogRequest/LogDatasets/LogDatasets/Dataset[@name="..."]

afishbeck replied:
Agreed.  Will change.

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -919,15 +964,29 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
+        if (servicePLTs || methodPLTs)
+        {
+            processServiceAndMethodTransforms(scriptContext, {servicePLTs, methodPLTs}, ESDLScriptCtxSection_LogData, nullptr);
+            scriptContext->toXML(temp, ESDLScriptCtxSection_PreESDLResponse);
+            logdata = temp.str();
```
@afishbeck I don't understand what this should accomplish. PreESDLResponse does not appear to be populated after processing the transforms. It results in logdata=nullptr, and then the LogDatasets element is not available to the LogAgents because it is set on line 1000 below.

Did you mean to serialize `ESDLScriptCtxSection_LogData` here? If so then the LogDatasets will be passed to the LogAgents correctly if we also make the spelling change on line 978.

If we intend to pass the contents of PreESDLResponse to the agents as the LogDatasets, then we'll need to concatenate the LogData section also, or we'll need to see why the PreESDLResponse is not populated in this case.

afishbeck replied:
Yes, typo should be ESDLScriptCtxSection_LogData section being serialized.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -115,10 +183,19 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+            targetContext->ensureLocation(xpath, true);
+            processChildren(scriptContext, targetContext, xpathContext); //bulid nodeset
+            xpathContext->addXpathVariable(m_name, xpath);
+        }
```
Oops forgot the return above.   Not meant to fall through.

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1454,18 +1529,22 @@ void EsdlServiceImpl::prepareFinalRequest(IEspContext &context,
-            processServiceAndMethodTransforms(scriptContext, {m_serviceLevelRequestTransform, methodCrt}, ESDLScriptCtxSection_ESDLRequest, ESDLScriptCtxSection_FinalRequest);
+        if (serviceCRTs || serviceRTs || methodCRTs || methodRTs)
+            processServiceAndMethodTransforms(scriptContext, {serviceCRTs, serviceRTs, methodCRTs, methodRTs}, ESDLScriptCtxSection_ESDLRequest, ESDLScriptCtxSection_FinalRequest);
         scriptContext->toXML(reqProcessed.clear(), ESDLScriptCtxSection_FinalRequest);
```
I think this should be part of the `if (serviceCRTs || serviceRTs || methodCRTs || methodRTs)` clause. Otherwise when there are no transforms (Legacy or BackendRequest) we will clear out the reqProcessed which is just about to be sent to the back end.

afishbeck replied:
agreed

asselitx commented on testing/unittests/esdltests.cpp:
```diff
@@ -1008,7 +1032,7 @@ constexpr const char * result = R"!!(<soap:Envelope xmlns:soap="http://schemas.x
     <Param name='section' select="'Friends'"/>
     <Param name='garbage' select="''"/>
-    <Param name='ForBuildListPath' select='/esdl_script_context/ESDLRequest/root/extra/Friends/Name'/><!--absolute path may change, highly frowned upon-->
+    <Param name='ForBuildListPath' select='/esdl_script_context/esdl_request/root/extra/Friends/Name'/>
```
testEsdlTransformAbsoluteSoapPath above will also need this change ESDLRequest -> esdl_request

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -130,51 +207,312 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+            m_traceName.set(stag.getValue("name"));
+        m_name.set(stag.getValue("name"));
+        m_section.set(stag.getValue("section"));
+        if (m_section.isEmpty())
```
it might have just been set on the line above.  this is basically saying the default is temporaries if stag "section" is not set.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -919,15 +964,29 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
+bool EsdlServiceImpl::handleResultLogging(IEspContext &espcontext, IEsdlScriptContext *scriptContext, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, IPropertyTree * reqcontext, IPropertyTree * request,const char *rawreq, const char * rawresp, const char * finalresp, const char * logdata)
 {
+    StringBuffer temp;
+    if (scriptContext)
+    {
+        IEsdlTransformSet *servicePLTs = m_transforms->queryMethodEntryPoint("", ESDLScriptEntryPoint_PreLogging);
+        IEsdlTransformSet *methodPLTs = m_transforms->queryMethodEntryPoint(mthdef.queryName(), ESDLScriptEntryPoint_PreLogging);
+
+        scriptContext->appendContent(ESDLScriptCtxSection_LogData, "LogDataset", logdata);
+
+        if (servicePLTs || methodPLTs)
+        {
+            processServiceAndMethodTransforms(scriptContext, {servicePLTs, methodPLTs}, ESDLScriptCtxSection_LogData, nullptr);
+            scriptContext->toXML(temp, ESDLScriptCtxSection_PreESDLResponse);
+            logdata = temp.str();
+        }
+    }
```
JIRA is broken right now.  Let's try and remember to create one for this.  Either of us can do it when we are adding more entry points (or before).

afishbeck commented on initfiles/examples/EsdlExample/esdl_binding_entrypoints.xml:
```diff
@@ -0,0 +1,147 @@
+              </es:content>
+            </es:http-post-xml>
+            <es:target xpath="soap:Body">
+            <es:target xpath="resp:RoxieEchoPersonInfoResponse">
+            <es:target xpath="resp:Results/resp:Result">
+            <es:target xpath="ds1:Dataset[@name='RoxieEchoPersonInfoResponse']">
```
ok will fix.

afishbeck commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +392,374 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+    {
+        if (isEmptyString(xpath))
+            return true;
+        if (strstr(xpath, "//")) //maybe in the future, scripting error rather than not found, required==false has no effect
+            throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: '//' not currently allowed '%s'", xpath);
```
Agreed, but still not this PR.

afishbeck commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -341,15 +348,28 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
     xmlXPathObjectPtr findVariable(const char *name, const char *ns_uri, CLibXpathScope *scope)
     {
+        xmlXPathObjectPtr obj = nullptr;
+        if (parent)
+        {
+            obj = parent->findVariable(name, ns_uri, scope);
+            if (obj)
+                return obj;
+        }
```
Parent was a bad name.  In our case this is source context with variables and target context without.  I've changed the variable name to primary which may be less confusing.  Don't want to dictate future uses by calling it source, but it's not parent either.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -130,51 +207,312 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
     }
 };
 
+class CEsdlTransformOperationHttpContentXml : public CEsdlTransformOperationWithChildren
```
After I've added a few more operations we can clean this stuff up.  I still want to see how things work out.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -130,51 +207,312 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+    }
+};
+
+interface IEsdlTransformOperationHttpHeader : public IInterface
```
If this doesn't get reused in the next few operations I may simplify.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -727,37 +1094,34 @@ class CEsdlTransformOperationConditional : public CEsdlTransformOperationWithChi
+            m_op = 'w';
+        else if (streq(op, "otherwise"))
+            m_op = 'o';
+        if (m_op!='o')
```
Should never get here without a recognized name, but I will add that for future proofing.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -797,31 +1161,61 @@ class CEsdlTransformOperationConditional : public CEsdlTransformOperationWithChi
+                if (streq(op, "when"))
+                    operations.append(*new CEsdlTransformOperationConditional(xpp, opTag, prefix));
+                else if (streq(op, "otherwise"))
+                    otherwise.setown(new CEsdlTransformOperationConditional(xpp, opTag, prefix));
```
Adding checking.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -834,62 +1228,243 @@ class CEsdlTransformOperationChoose : public CEsdlTransformOperationWithChildren
+    virtual ~CEsdlTransformOperationEnsureTarget(){}
+};
+
+class CEsdlTransformOperationSource : public CEsdlTransformOperationWithChildren
```
Can refactor later.  I'd like to wait a bit in case these diverge.

afishbeck commented:
@timothyklemm @asselitx updated, please review.  I'll have to resolve some new merge conflicts soon.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 31d4f2827d5a33704d4e439129b7dc207e7b989e
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-219.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-219.ca-central-1.compute.internal
Sha: 31d4f2827d5a33704d4e439129b7dc207e7b989e
Conflicting files, should skip build and test.

[Build error:](https://gist.github.com/6f0400b2474573ed1ff6539731f8dd44)
Build log [Build log](https://gist.github.com/HPCCSmoketest/6f0400b2474573ed1ff6539731f8dd44#file-result-21-01-05-22-58-25-log)

timothyklemm commented on esp/bindings/SOAP/xpp/xpp/XmlPullParser.h:
```diff
@@ -174,7 +174,8 @@ namespace xpp {
 
 
-    int skipSubTree()  {
+    bool skipSubTree()  {
```
The platform isn't currently using the return value, so this change looks reasonable. But should we assume that nothing else uses the existing return value?

afishbeck replied:
The code couldn't make any real use of it since it always returned XmlPullParser::END_TAG whether it cycled through some children or not, but I'll make a new function anyway.

timothyklemm replied:
Your change is fine to address the original comment, but I am concerned about the original implementation.

If skipSubTree is used on a document not known to be well-formed, an unbalanced start tag can put it into an infinite loop. It only leaves the loop when the end tags balance. I remember writing code before where I checked to make sure I wasn't at the end of the document after skipping. I didn't realize until now that it's an unhandled error, unless I've misread next() and it doesn't return END_DOCUMENT infinitely.

It's also not clear that the well-formed behavior is consistent given different starting positions. Skipping from the start tag skips the entire element. Skipping from content skips the remainder of the current element. Skipping from the end tag of the current element skips the remainder of the parent element (and foolishly skipping from the end of the root element will never return). The first two make sense to me, but I could easily see the third being a NO-OP instead.

timothyklemm reviewed (commented):
@afishbeck Just a few questions/comments.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -1407,7 +1409,7 @@ void createEsdlTransformOperations(IArrayOf<IEsdlTransformOperation> &operations
 }
 
-IEsdlTransformOperation *createEsdlTransformOperation(XmlPullParser &xpp, const StringBuffer &prefix, bool withVariables)
+IEsdlTransformOperation *createEsdlTransformOperation(XmlPullParser &xpp, const StringBuffer &prefix, bool withVariables, bool ignoreCodingErrors)
```
The parameter has to be there because the function has to match the signature "esdlOperationsFactory_t" which allows functions to make use of the parameter, but doesn't require it.  It would generally not be passed to a given operation since they have their own ignore option.  Rather if the function itself is being called as a child and has an error in it's own code, it needs to know what to do.

afishbeck commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -718,10 +741,10 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>, implements IVariabl
         }
         xmlNodePtr *nodes = obj->nodesetval->nodeTab;
-        for (int i = 0; i < obj->nodesetval->nodeNr; i++)
+        for (int i = obj->nodesetval->nodeNr; i >= 0; i--)
```
Yes, thanks for catching that.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1526,10 +1519,10 @@ void EsdlServiceImpl::prepareFinalRequest(IEspContext &context,
-    StringAttr *crtErrorMessage = m_requestTransformErrors.getValue(mthName);
+    StringAttr *crtErrorMessage = m_methodScriptErrors.getValue(mthName);
     if (crtErrorMessage && !crtErrorMessage->isEmpty())
         throw MakeStringException(-1, "%s::%s disabled due to method-level Custom Transform errors: %s. Review transform template in configuration.", srvdef.queryName(), mthName, crtErrorMessage->str());
```
I'll try and make consistent.

afishbeck commented:
@timothyklemm @asselitx I've updated based on comments, but also rebased to latest master.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 83f2852a422eff7943029f73069583a0559e9d08
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-180.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-180.ca-central-1.compute.internal
Sha: 83f2852a422eff7943029f73069583a0559e9d08
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |24 sec  (00:00:24) |
|setup (thor) |9 |9 |0 |39 sec  (00:00:39) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |947 |947 |0 |326 sec  (00:05:26) |
|test (thor) |858 |858 |0 |775 sec  (00:12:55) |
|test (roxie) |1020 |1020 |0 |351 sec  (00:05:51) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  591 sec (00:09:51) |  104 sec (00:01:44) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1762 sec (00:29:22) |   18 sec (00:00:18) | 2520 sec (00:42:00) |

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -69,44 +85,89 @@ static inline StringBuffer &makeOperationTagName(StringBuffer &s, const StringBu
         if (m_traceName.isEmpty())
-            m_traceName.set(tree->queryProp("@name"));
-        m_name.set(tree->queryProp("@name"));
+            m_traceName.set(stag.getValue("name"));
```
Might be beneficial to have this logic in the base class to ensure that all operations default to the name as traceName when its empty.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -115,11 +176,22 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
-            return xpathContext->addCompiledVariable(m_name, m_select);
-        return xpathContext->addVariable(m_name, "");
+            return sourceContext->addCompiledVariable(m_name, m_select);
+        else if (m_children.length())
```
This case assigns the value of the variable to the result of processing the children?

afishbeck replied:
Yes, exactly.  It creates a temporary location for that xml in a special section of the script context tree and then changes the target for processing of the children to that location.  The end result is structured xml content stored in the variable.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -130,51 +202,312 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+            {
+                throw MakeStringException(ESDL_SCRIPT_Error, "ESDL Script empty result calling http-post-xml %s url(%s)", m_traceName.str(), url);
+            }
+            else if (*response.str()=='<')
```
I don't know if we really can or need to do anything differently here, but does the HTTP spec say anything about guarantees of response type given a certain Content Type? I'm guessing not really, but we could have a situation here where some text is returned and just happens to start with '<'. Then we'd just return whatever XML parse error resulting from addXmlContent().

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -328,42 +666,86 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+    CEsdlTransformOperationNamespace(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationWithoutChildren(xpp, stag, prefix)
+    {
+        if (m_traceName.isEmpty())
+            m_traceName.set(stag.getValue("name"));
```
Does the namespace operation really use @name? If not perhaps @prefix or @uri would be better fallbacks for the traceName.

afishbeck replied:
Yeah even the original use of name was a bit weird.  I'll remove it or use one of the other values for now.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -328,42 +666,86 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+
+        const char *pfx = stag.getValue("prefix");
+        const char *uri = stag.getValue("uri");
+        if (!pfx && isEmptyString(uri))
```
Not a big deal at all but why not use isEmptyString() to check both?

afishbeck replied:
"" is a valid prefix, but if you are relying on the uri it can't be empty.  That is , pfx=nullptr, uri="" has nothing to match on, pfx="", uri=nullprt can search for an existing prefix definition for "".

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -328,42 +666,86 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
-        if (isEmptyString(tree->queryProp("@new_name")) && isEmptyString(tree->queryProp("@xpath_new_name")))
+        const char *new_name = stag.getValue("new_name");
+        const char *xpath_new_name = stag.getValue("xpath_new_name");
+        if (isEmptyString(new_name) && isEmptyString(xpath_new_name))
```
Perhaps set m_traceName to new_name or xpath_new_name if empty?

afishbeck replied:
will do.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -377,29 +759,75 @@ class CEsdlTransformOperationRenameNode : public CEsdlTransformOperationBase
             else
                 name.set(m_new_name);
 
             //make sure we only rename one.  we can add a multi-node support in the future (when we are libxml2 based?)
```
Is this comment now obsolete?

afishbeck replied:
yes, I think so.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -834,62 +1230,243 @@ class CEsdlTransformOperationChoose : public CEsdlTransformOperationWithChildren
-        m_children.append(*new CEsdlTransformOperationConditional(otherwise, prefix));
+        CXpathContextLocation location(targetContext);
+        bool success = false;
+        if (m_ensure)
```
Will we be able to set this to true in the future via an `@ensure` attribute?

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -834,62 +1230,243 @@ class CEsdlTransformOperationChoose : public CEsdlTransformOperationWithChildren
+
+public:
+    CEsdlTransformOperationElement(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationWithChildren(xpp, stag, prefix, true, nullptr)
+    {
```
if `@trace` is empty could we fallback to using `@name`?

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -1024,7 +1541,7 @@ class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
             prefixes.appendUniq("");
     }
-    virtual void toDBGLog() override
+    virtual void toDBGLogx() override
```
Is the trailing x in the function name a typo?

afishbeck replied:
oops that was me tracking down usage.  I'll change the name back.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+
+        CLibCompiledXpath * ccXpath = static_cast<CLibCompiledXpath *>(compiledXpath);
+        if (!ccXpath)
+            throw MakeStringException(XPATHERR_InvalidState,"XpathProcessor:evaluateAsNumber: Error: BAD compiled XPATH");
```
Shouldn't the error read `XpathProcessor:setLocation`?

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+    virtual bool setLocation(const char *s, bool required) override
+    {
+        StringBuffer xpath;
+        replaceVariables(xpath, s, false, this, "{$", "}");
```
So this change allows us to use the ${...} variable substitution as strings in xpaths wherever this function is used?

afishbeck replied:
Yes let me know when you'd like to discuss it.  I'll have to refresh my memory on exactly where this type of substitution is supported.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+
+            if (obj->type!=xmlXPathObjectType::XPATH_NODESET || !obj->nodesetval || obj->nodesetval->nodeNr==0)
+            {
+                if (*finger=='.' || strpbrk(finger, "[()]*") || strstr(finger, "::")) //complex nodes must already exist
```
Will be good to ensure that we document this constraint.

afishbeck replied:
Yes.  Once explained in the docs it would be pretty clear why this is necessary.  For example I can't add a node named a*b, but I could fine a node that matched that pattern.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+        if (isEmptyString(uri))
+            ns = xmlSearchNs(node->doc, node, (const xmlChar *) prefix);
+        else
+            ns = xmlNewNs(node, (const xmlChar *) uri, (const xmlChar *) prefix);
```
What if the prefix/uri is a redefinition? Will xmlNewNs throw an error? Or what if the prefix is reused with a different uri? Asking mostly because some of the libxml errors aren't very friendly in that they don't  show up in the esp.log file.

afishbeck replied:
Right, I don't think you'd get an error redefining a prefix.  And routines relying on it would find the nearest location where the prefix was defined.  libxml errors and implementation can be a bit quirky though.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -195,14 +195,16 @@ class XpathContextState
     XPathInputMap provided;
     xmlDocPtr m_xmlDoc = nullptr;
     xmlXPathContextPtr m_xpathContext = nullptr;
     ReadWriteLock m_rwlock;
```
Which operations does this need to protect? Only some of the methods in this class use it. Some do a fair amount of work before calling `setLocation()` and others like `rename()` and `remove()` don't appear to use it at all, which seems like it might be an oversight.

afishbeck replied:
I may just remove this in the future.  The scripting is not multithreaded and would have to be majorly refactored to make it so.  It's not a declarative language.

asselitx commented:
I've completed my review @afishbeck . With the possible exception of the use of m_rwlock in libxml_xpathprocessor.cpp none of my comments are critical - at worst they may be future cleanup.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -410,74 +410,155 @@ void EsdlServiceImpl::configureUrlMethod(const char *method, IPropertyTree &entr
+    raise.append("xsdl:variable");
+
+    StringArray lower;
+    lower.append("xsdl:otherwise");
```
just here for completeness, and to make the modified xml read more accurately.

afishbeck commented on system/xmllib/xpathprocessor.hpp:
```diff
@@ -90,23 +105,45 @@ class CXpathContextScope : CInterface
+    Linked<IXpathContext> context;
+public:
+    IMPLEMENT_IINTERFACE;
+    CXpathContextLocation(IXpathContext *ctx) : context(ctx)
```
I now check for null.  I want this to be easy to use and not have anyone worry about the current value, casting to reference, etc.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+        {
+            const char *finger = xpathnodes.item(i);
+            if (isEmptyString(finger)) //shouldn't happen, scripting error rather than not found, required==false has no effect
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureValue: '//' not allowed '%s'", xpath);
```
Possibly misleading message. You've already rejected any occurrence of "//", so this is caused by trimmed whitespace when splitting the path, or (possibly) a leading or trailing '/'. It would be better to distinguish this exception message from the one above.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -1064,6 +1499,36 @@ class CEsdlScriptContext : public CInterfaceOf<IEsdlScriptContext>
+
+    xmlNodePtr ensureCopiedSection(const char *target, const char *source, bool replace)
+    {
+        sanityCheckSectionName(target);
+        sanityCheckSectionName(source);
+        if (isEmptyString(source)||isEmptyString(target))
+            return nullptr;
```
These checks may be backwards. If strpbrk, used by sanityCheckSectionName, is anything like other standard library functions that behave badly when passed null, null should be checked first (or as part of the sanity check if it makes sense).

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1431,22 +1505,31 @@ void EsdlServiceImpl::prepareFinalRequest(IEspContext &context,
     if (scriptContext)
     {
-        IEsdlCustomTransform *methodCrt = m_customRequestTransformMap.getValue(mthName);
+        IEsdlTransformSet *serviceCRTs = m_transforms->queryMethodEntryPoint("", ESDLScriptEntryPoint_Legacy);
+        IEsdlTransformSet *methodCRTs = m_transforms->queryMethodEntryPoint(mthName, ESDLScriptEntryPoint_Legacy);
+
+        IEsdlTransformSet *serviceRTs = m_transforms->queryMethodEntryPoint("", ESDLScriptEntryPoint_BackendRequest);
+        IEsdlTransformSet *methodRTs = m_transforms->queryMethodEntryPoint(mthName, ESDLScriptEntryPoint_BackendRequest);
```
These aren't unordered and ordered.  These are entry point names.  They are equavent to each other and if I add a "BackEndRequest" entry point to a service that already has Legacy entry points it will work.  Whether things are ordered or now is based on how the xml is added as CDATA or inline.

timothyklemm replied:
I get the technicalities, but in reality there is no existing legacy markup that uses CDATA (thus all such markup is unordered) and new markup will likely use CDATA (and be ordered). I question whether it is necessary and whether an arbitrary decision that legacy transforms will be processed first will be sufficient. Maybe I (as a potential end user) want new markup to set things up for the legacy, so should we support an option to switch the order? Maybe I have an unrealistic expectation that I should somehow be able to intermingle old and new.

My thought (as a reviewer) is that we should support both, but not together. It's just my opinion, and I'm probably still influenced by my abandoned refactoring where I wasn't even looking for legacy if new markup was found. I also wasn't allowing legacy syntax to include new operations. It was pick one or the other, but never both.

afishbeck replied:
We will eventually deprecate the old name, but I don't want to dictate that people have to do that right away just to use a new tag.  I'm going to fix it so that the order will be preserved across both types.

afishbeck commented:
@asselitx @timothyklemm updated based on comments.  Please check httpclient changes (required to check content-type) for any possible regressions as its a shared library.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 8d0ce97871adb7b89d86f9bb4796c0ff918232af
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-42.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-42.ca-central-1.compute.internal
Sha: 8d0ce97871adb7b89d86f9bb4796c0ff918232af
Number of Error(s): 9
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/protocols/http/./../../bindings/http/platform/httptransport.ipp:165:27: error: \xe2\x80\x98virtual StringBuffer& CHttpMessage::getHeader(const char*, StringBuffer&)\xe2\x80\x99 marked \xe2\x80\x98override\xe2\x80\x99, but does not override
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/protocols/http/./../../bindings/http/platform/httptransport.ipp:282:27: error: \xe2\x80\x98virtual StringBuffer& CHttpMessage::getStatus(StringBuffer&)\xe2\x80\x99 marked \xe2\x80\x98override\xe2\x80\x99, but does not override
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/bindings/http/client/httpclient.cpp:771:11: error: \xe2\x80\x98struct IHttpMessage\xe2\x80\x99 has no member named \xe2\x80\x98getStatus\xe2\x80\x99
make[2]: *** [esp/protocols/http/CMakeFiles/esphttp.dir/build.make:76: esp/protocols/http/CMakeFiles/esphttp.dir/__/__/bindings/http/client/httpclient.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10600: esp/protocols/http/CMakeFiles/esphttp.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-01-19_16-31-54
time logs started
Prep time   :    10 sec (00:00:10)
Build time  :   431 sec (00:07:11)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   441 sec (00:07:21)
time logs end

[Build error:](https://gist.github.com/8890414f64c6d82d8650a4ef0c1f9874)
Build log [Build log](https://gist.github.com/HPCCSmoketest/8890414f64c6d82d8650a4ef0c1f9874#file-result-21-01-19-16-24-21-log)

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 95ffdb79a430f0a2bd0550b647f3b21c10698ea7
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal
Sha: 95ffdb79a430f0a2bd0550b647f3b21c10698ea7
Number of Error(s): 9
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/protocols/http/./../../bindings/http/platform/httptransport.ipp:165:27: error: \xe2\x80\x98virtual StringBuffer& CHttpMessage::getHeader(const char*, StringBuffer&)\xe2\x80\x99 marked \xe2\x80\x98override\xe2\x80\x99, but does not override
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/protocols/http/./../../bindings/http/platform/httptransport.ipp:282:27: error: \xe2\x80\x98virtual StringBuffer& CHttpMessage::getStatus(StringBuffer&)\xe2\x80\x99 marked \xe2\x80\x98override\xe2\x80\x99, but does not override
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/bindings/http/client/httpclient.cpp:771:11: error: \xe2\x80\x98struct IHttpMessage\xe2\x80\x99 has no member named \xe2\x80\x98getStatus\xe2\x80\x99
make[2]: *** [esp/protocols/http/CMakeFiles/esphttp.dir/build.make:76: esp/protocols/http/CMakeFiles/esphttp.dir/__/__/bindings/http/client/httpclient.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10600: esp/protocols/http/CMakeFiles/esphttp.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-01-19_18-05-02
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   440 sec (00:07:20)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   448 sec (00:07:28)
time logs end

[Build error:](https://gist.github.com/95b024da6ce4d2ae3f148f9ef0429474)
Build log [Build log](https://gist.github.com/HPCCSmoketest/95b024da6ce4d2ae3f148f9ef0429474#file-result-21-01-19-17-57-21-log)

asselitx commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -345,6 +345,67 @@ static void storedValueExistsFunction (xmlXPathParserContextPtr ctxt, int nargs)
+    if (!scriptContext->tokenize((const char *)str, (const char *)delimiters, resultPath))
+    {
+        xmlXPathSetError((ctxt), XPATH_EXPR_ERROR);
+        return;
```
Also what about str? Could it be leaked here or above when `if(xmlXPathCheckError(ctxt)...)` is true and str!=NULL?

afishbeck replied:
done.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -396,22 +396,37 @@ class CEsdlTransformOperationHttpPostXml : public CEsdlTransformOperationBase
                 targetContext->addElementToLocation("content");
                 targetContext->addXmlContent(response.str());
             }
             else
```
Curious if, with the availability of content-type, it might be beneficial to recognize JSON (or anything we can load into a PTree) then convert to XML. Add the text so somebody expecting the actual response can get it, but do the conversion automatically so it's ready to use.

On the other hand, if any non-XML content type is considered an error, should we make sure the request asks for XML results?

afishbeck replied:
I'll ask for xml.  I named the function with xml in the name to put off the complexity of conversion.  If I were already loading into a ptree it would be different.  Later I can probably create something that parses json directly into libxml.  I don't like the ptree as an intermediary here.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -417,17 +845,18 @@ class CEsdlTransformOperationRenameNode : public CEsdlTransformOperationBase
 protected:
     StringAttr m_target;
     Owned<ICompiledXpath> m_xpath_target;
+    bool all = false;
```
Only one of two operations with this comment has been changed.

asselitx commented:
@afishbeck httpclient changes look good

richardkchapman commented:
Some build failures of the form:

D:\a\HPCC-Platform\HPCC-Platform\esp\bindings\http/platform/httptransport.ipp(165,27): error C3668: 'CHttpMessage::getHeader': method with override specifier 'override' did not override any base class methods [D:\a\HPCC-Platform\HPCC-Platform\build\esp\protocols\http\esphttp.vcxproj]
1333

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 296e5f5e9314cb3082d59ec2e888d59d99e55495
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-89.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-89.ca-central-1.compute.internal
Sha: 296e5f5e9314cb3082d59ec2e888d59d99e55495
Number of Error(s): 7
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-14331/HPCC-Platform/esp/bindings/http/client/httpclient.cpp:771:11: error: \xe2\x80\x98struct IHttpMessage\xe2\x80\x99 has no member named \xe2\x80\x98getStatus\xe2\x80\x99
make[2]: *** [esp/protocols/http/CMakeFiles/esphttp.dir/build.make:76: esp/protocols/http/CMakeFiles/esphttp.dir/__/__/bindings/http/client/httpclient.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10600: esp/protocols/http/CMakeFiles/esphttp.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-01-21_01-01-55
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   436 sec (00:07:16)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   444 sec (00:07:24)
time logs end

[Build error:](https://gist.github.com/c30a67ef34db3bb57833d48b29369a83)
Build log [Build log](https://gist.github.com/HPCCSmoketest/c30a67ef34db3bb57833d48b29369a83#file-result-21-01-21-00-54-19-log)

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: a693c11245d3c7d82cea3e9076dfa270d8f2c172
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-145.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-145.ca-central-1.compute.internal
Sha: a693c11245d3c7d82cea3e9076dfa270d8f2c172
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |26 sec  (00:00:26) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |20 sec  (00:00:20) |
|test (hthor) |947 |947 |0 |647 sec  (00:10:47) |
|test (thor) |858 |858 |0 |1073 sec  (00:17:53) |
|test (roxie) |1020 |1020 |0 |766 sec  (00:12:46) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  612 sec (00:10:12) |  108 sec (00:01:48) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1383 sec (00:23:03) |   18 sec (00:00:18) | 2165 sec (00:36:05) |

afishbeck commented:
Fixed build issue.
Also added sending an Accept header from the post-xml script method.
@asselitx @timothyklemm please review.  Hopefully the last time.  Then I'll squash and rebase.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 2c2255a8305c613b4bdec7f3f4e73e14123feb45
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-152.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-152.ca-central-1.compute.internal
Sha: 2c2255a8305c613b4bdec7f3f4e73e14123feb45
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |42 sec  (00:00:42) |
|setup (roxie) |19 |19 |0 |19 sec  (00:00:19) |
|test (hthor) |947 |946 |1 |1827 sec  (00:30:27) |
|test (thor) |858 |858 |0 |1093 sec  (00:18:13) |
|test (roxie) |1020 |1020 |0 |733 sec  (00:12:13) |


[Errors:](https://gist.github.com/7bb9e62ec7bffa8638f288fd2dd9bd1f)
- test
  - hthor
    - 604\. [schedule1](https://gist.github.com/HPCCSmoketest/7bb9e62ec7bffa8638f288fd2dd9bd1f#file-schedule1-w20210121-163814-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  599 sec (00:09:59) |  106 sec (00:01:46) |   20 sec (00:00:20) |   16 sec (00:00:16) | 2134 sec (00:35:34) |   17 sec (00:00:17) | 2900 sec (00:48:20) |

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -727,37 +1089,37 @@ class CEsdlTransformOperationConditional : public CEsdlTransformOperationWithChi
     virtual ~CEsdlTransformOperationConditional(){}
 
-    bool process(IEsdlScriptContext * context, IPropertyTree *content, IXpathContext * xpathContext) override
+    virtual bool process(IEsdlScriptContext * scriptContext, IXpathContext * targetContext, IXpathContext * sourceContext) override
```
Who did you have the debate with?  I saw a comment somewhere else that it is stylistically preferred in this project.. so following that recommendation.. and personally prefer it this way.  If there was another consensus we can all follow it, I just don't know of one.

afishbeck commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -903,118 +1480,58 @@ static inline void replaceVariable(StringBuffer &s, IXpathContext *xpathContext,
 private:
-    IArrayOf<IEsdlTransformOperation> m_variables; //keep separate and only at top level for now
     IArrayOf<IEsdlTransformOperation> m_operations;
     Owned<IProperties> namespaces = createProperties(false);
```
This PR is to big for genaral cleanup of existing code.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -899,15 +943,29 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
+bool EsdlServiceImpl::handleResultLogging(IEspContext &espcontext, IEsdlScriptContext *scriptContext, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, IPropertyTree * reqcontext, IPropertyTree * request,const char *rawreq, const char * rawresp, const char * finalresp, const char * logdata)
 {
+    StringBuffer temp;
+    if (scriptContext)
```
This is conditional on the scripts having been configured.  It's up to the person creating the binding.  Despite the name these scripts could be used for anything.  Calling out to another service at completion for example.
Open another JIRA about TxSummary entries for scripts.

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 7e4d722aaf7844d61f5e8544d2aeb3d9972931a9
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-84.ca-central-1.compute.internal

afishbeck commented:
@asselitx @timothyklemm this needs final sign off.  I saw a couple more comments I missed and made a couple changes.  Please review and sign off.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-84.ca-central-1.compute.internal
Sha: 7e4d722aaf7844d61f5e8544d2aeb3d9972931a9
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |20 sec  (00:00:20) |
|test (hthor) |951 |951 |0 |671 sec  (00:11:11) |
|test (thor) |862 |862 |0 |1088 sec  (00:18:08) |
|test (roxie) |1024 |1024 |0 |739 sec  (00:12:19) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  597 sec (00:09:57) |  105 sec (00:01:45) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1393 sec (00:23:13) |   18 sec (00:00:18) | 2157 sec (00:35:57) |

asselitx reviewed (dismissed):
Looks good

HPCCSmoketest commented:
Process of PR-14331, label: esdlScriptNovUpdate2 starts now.
The reason of this test is: Forced to re-test.
Commit ID: a1211e17c58bee5de2615db305fd8d39e0d6a5aa
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal
Sha: a1211e17c58bee5de2615db305fd8d39e0d6a5aa
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |26 sec  (00:00:26) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |953 |953 |0 |657 sec  (00:10:57) |
|test (thor) |864 |864 |0 |1131 sec  (00:18:51) |
|test (roxie) |1026 |1026 |0 |770 sec  (00:12:50) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  606 sec (00:10:06) |  107 sec (00:01:47) |   20 sec (00:00:20) |   20 sec (00:00:20) | 1439 sec (00:23:59) |   17 sec (00:00:17) | 2217 sec (00:36:57) |

afishbeck commented:
@richardkchapman rebased and squashed.  This PR should be ready to merge.
